File: ex5-regularization/ex5.m
x = load('ex5Linx.dat');
y = load('ex5Liny.dat');
[m, n] = size(x);
x = [ones(m, 1), x, x.^2, x.^3, x.^4, x.^5];
[m, n] = size(x);


L = diag([0, ones(1,n-1)]);

theta_0 = (x' * x) \ (x' * y);
theta_1 = (x' * x + L) \ (x' * y);
theta_10 = (x' * x + 10 * L) \ (x' * y);

xs = linspace(-1, 1, 100)';
xs = [xs.^0, xs, xs.^2, xs.^3, xs.^4, xs.^5];

figure;
hold on;
plot(x(:,2),y, 'o');
plot(xs(:,2), xs * theta_0, 'r--')
plot(xs(:,2), xs * theta_1, 'g--')
plot(xs(:,2), xs * theta_10, 'b--')
legend('Training data', 'lambda=0', 'lambda=1', 'lambda=10')
hold off;


% Logistic Regression %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
clear all; close all;

x = load('ex5Logx.dat');
u = x(:,1); v = x(:,2);
y = load('ex5Logy.dat');

g = inline('1.0 ./ (1.0 + exp(-z))')

X = map_feature(u, v);
[m, n] = size(X);

eye_0 = diag([0; ones(n-1,1)]);

lambdas = [0, 1, 10];

thetas = [];
for i = 1:length(lambdas)
  theta = zeros(n, 1);
  lambda = lambdas(i);
  Js = [];
  diff = ones(n, 1);
  while norm(diff) > 1e-6
    h = g(X * theta);
    theta_0 = [0; theta(2:end,1)];
    J = -(1/m) * sum(y.*log(h) + (1-y).*log(1 - h)) + (lambda/2*m) * theta_0' * theta_0;
    Js = [Js, J];
    grad = ((1/m) * X' * (h - y)) + ((lambda/m) * theta_0);
    H = ((1/m) * (repmat(h .* (1 - h), 1, n) .* X)' * X) + ((lambda/m) * eye_0);
    diff = H \ grad;
    theta = theta - diff;
  end
  iterations = length(Js)
  thetas = [thetas, theta];
end

% Define the ranges of the grid
uu = linspace(-1, 1.5, 200);
vv = linspace(-1, 1.5, 200);

% Initialize space for the values to be plotted
z0 = zeros(length(uu), length(vv));
z1 = zeros(length(uu), length(vv));
z10 = zeros(length(uu), length(vv));

% Evaluate z = theta*x over the grid
for i = 1:length(uu)
    for j = 1:length(vv)
        % Notice the order of j, i here!
        z0(j,i) = map_feature(uu(i), vv(j))*thetas(:,1);
        z1(j,i) = map_feature(uu(i), vv(j))*thetas(:,2);
        z10(j,i) = map_feature(uu(i), vv(j))*thetas(:,3);
    end
end

% Because of the way that contour plotting works
% in Matlab, we need to transpose z, or
% else the axis orientation will be flipped!
%
% But not in octave...

if (exist ('octave_config_info') ~= 0)
  % this is octave
else
  % this is not octave
  z0 = z0';
  z1 = z1';
  z10 = z10';
end

figure;
hold on;
plot(x(y==0,1), x(y==0,2), 'o');
plot(x(y==1,1), x(y==1,2), '+');

% Plot z = 0 by specifying the range [0, 0]
contour(uu,vv,z0, [0, 0], 'LineWidth', 2, 'r-');
contour(uu,vv,z1, [0, 0], 'LineWidth', 2, 'g-');
contour(uu,vv,z10, [0, 0], 'LineWidth', 2, 'b-');

xlabel('u'); ylabel('v');

legend('y = 0', 'y = 1', 'lambda=0', 'lambda=1', 'lambda=10');

File: ex5-regularization/ex5.py
#! ~/.ve/np/bin/python

import time
import sys
import numpy as np
import scipy as sp
import matplotlib as mpl
if sys.platform == 'darwin': mpl.use('TkAgg')

import matplotlib.pyplot as plt
from scipy import linalg
from mpl_toolkits.mplot3d import Axes3D

x = np.loadtxt('ex5Linx.dat')[:, np.newaxis]
y = np.loadtxt('ex5Liny.dat')[:, np.newaxis]

def f1(x):
  return np.hstack([x**0, x, x**2, x**3, x**4, x**5])

x = f1(x)
m, n = x.shape
L = np.diag(np.ones(n)); L[0, 0] = 0

theta_0 = linalg.solve(x.T.dot(x), x.T.dot(y))
theta_1 = linalg.solve(x.T.dot(x) + L, x.T.dot(y))
theta_10 = linalg.solve(x.T.dot(x) + 10 * L, x.T.dot(y))

xs = np.linspace(-1, 1, 100)[:, np.newaxis]
xs = f1(xs)

plt.figure()
p1 = plt.plot(x[:, 1:2], y, 'o')
p2 = plt.plot(xs[:, 1:2], xs.dot(theta_0), 'r--')
p3 = plt.plot(xs[:, 1:2], xs.dot(theta_1), 'g--')
p4 = plt.plot(xs[:, 1:2], xs.dot(theta_10), 'b--')
plt.legend((p1, p2, p3, p4), ('Training data', '$\lambda=0$', '$\lambda=1$', '$\lambda=10$'), 'upper right')



# Logistic Regression ########################################################

x = np.loadtxt('ex5Logx.dat', delimiter=',')
y = np.loadtxt('ex5Logy.dat')[:, np.newaxis]
u = x[:, 0:1]; v = x[:, 1:2]

def g(z):
    return 1.0 / (1.0 + np.exp(-z))

def map_feature(u, v):
  if isinstance(u, (int, long, float)):
    u = np.array([[u]])
    v = np.array([[v]])
  ret = np.empty([u.size, 0])
  degree = 6
  for i in range(0, degree+1):
    for j in range(0, degree-i+1):
      ret = np.hstack([ret, u**i * v**j])
  return ret

X = map_feature(u, v)
m, n = X.shape

eye_0 = np.eye(n); eye_0[0, 0] = 0

lambdas = [0.0, 1.0, 10.0]

thetas = np.empty([n, 0])
for lam in lambdas:
  theta = np.zeros([n,1])
  diff = np.ones([n,1])
  iterations = 0
  while linalg.norm(diff) > 1e-6:
    iterations += 1
    h = g(X.dot(theta))
    theta_0 = eye_0.dot(theta)
    J = -(1.0/m) * np.sum(y * np.log(h) + (1-y) * np.log(1-h)) + (lam/(2*m)) * theta_0.T.dot(theta_0)
    grad = (1.0/m) * X.T.dot(h-y) + ((lam/m) * theta_0)
    H = (1.0/m) * (h * (1-h) * X).T.dot(X) + ((lam/m) * eye_0)
    diff = linalg.solve(H, grad)
    theta = theta - diff
  print "Iterations: %d" % iterations
  thetas = np.hstack([thetas, theta])

# Plotting contours

grid_size = 200
uu = np.linspace(-1, 1.5, grid_size)[:, np.newaxis]
vv = np.linspace(-1, 1.5, grid_size)

z0 = np.empty([grid_size, 0])
z1 = np.empty([grid_size, 0])
z10= np.empty([grid_size, 0])

for vvj in vv:
  z0 = np.hstack([z0, map_feature(uu, np.array([[vvj]])).dot(thetas[:, 0:1])])
  z1 = np.hstack([z1, map_feature(uu, np.array([[vvj]])).dot(thetas[:, 1:2])])
  z10 = np.hstack([z10, map_feature(uu, np.array([[vvj]])).dot(thetas[:, 2:3])])

z0 = z0.T
z1 = z1.T
z10 = z10.T

plt.figure()
p1 = plt.plot(x[:,0:1][y==1], x[:,1:2][y==1], '+')
p2 = plt.plot(x[:,0:1][y==0], x[:,1:2][y==0], 'o', mec='blue', mfc='none')
p3 = plt.contour(uu[:, 0], vv, z0, [0], colors='r')
p4 = plt.contour(uu[:, 0], vv, z1, [0], colors='g')
p5 = plt.contour(uu[:, 0], vv, z10, [0], colors='b')
plt.xlabel('u')
plt.ylabel('v')
p1[0].set_label('y=1')
p2[0].set_label('y=0')
p3.collections[0].set_label(r'$\lambda=0$')
p4.collections[0].set_label(r'$\lambda=1$')
p5.collections[0].set_label(r'$\lambda=10$')
plt.legend()
plt.show()
